### PR TITLE
Reset customer password bug. Fixed retrieving first item from Search Results

### LINKED ISSUE #29780: [Issue] Reset customer password bug. Fixed retrieving first item from Search Results
This issue is automatically created based on existing pull request: magento/magento2#29770: Reset customer password bug. Fixed retrieving first item from Search Results

---------

This PR fixes retrieving of first item from the Search Results Items executed during forgot password processing.

### Description (*)
In the logs I have a notice:
`Message: Notice: Undefined offset: 0 in /var/www/app/magento/vendor/magento/module-customer/Model/ForgotPasswordToken/GetCustomerByToken.php on line 87`

Magento collections/search results items are indexed by their ids, not by consecutive numbers. 

### Manual testing scenarios (*)

I've only seen error entries in a logs. Our customers complaining that they were unable reset their password.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by MTheProgrammer**
This PR fixes retrieving of first item from the Search Results Items executed during forgot password processing.

### Description (*)
In the logs I have a notice:
`Message: Notice: Undefined offset: 0 in /var/www/app/magento/vendor/magento/module-customer/Model/ForgotPasswordToken/GetCustomerByToken.php on line 87`

Magento collections/search results items are indexed by their ids, not by consecutive numbers. 

### Manual testing scenarios (*)

I've only seen error entries in a logs. Our customers complaining that they were unable reset their password.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29780: Reset customer password bug. Fixed retrieving first item from Search Results

### CODE REVIEW

**Comment by rogyar** [comment]
Hi @MTheProgrammer. Thank you for your collaboration, this fix might help in case if, for some reason, the indexes in the result array (`$found->getItems()`) are not being started from "0". 
Could I ask you to cover this case with a unit test? So, the test will mock the situation when `$found->getItems()` returns something like 

```php
[
    3 => Object1,
    5 => Object2
]
```

and assert that we still have the first element as the result.

Thank you.

**Comment by sidolov** [comment]
@magento create issue

**Comment by MTheProgrammer** [comment]
> Hi @MTheProgrammer. Thank you for your collaboration, this fix might help in case if, for some reason, the indexes in the result array (`$found->getItems()`) are not being started from "0".
> Could I ask you to cover this case with a unit test? So, the test will mock the situation when `$found->getItems()` returns something like
> 
> ```
> [
>     3 => Object1,
>     5 => Object2
> ]
> ```
> 
> and assert that we still have the first element as the result.
> 
> Thank you.

There is a guard with exception thrown inside it's body so it is always going to be just one item in the array.
`if ($found->getTotalCount() > 1) {`
Thus I'll create test against array with one item which has a key greater than 0.

**Comment by rogyar** [comment]
@magento run all tests

**Comment by MTheProgrammer** [comment]
@magento run all tests

**Comment by MTheProgrammer** [comment]
@magento run all tests

**Comment by MTheProgrammer** [comment]
@magento run Functional Tests B2B, Integration Tests

File: app/code/Magento/Customer/Model/ForgotPasswordToken/GetCustomerByToken.php [hunk 1]
```diff
@@ -84,6 +84,8 @@ public function execute(string $resetPasswordToken):CustomerInterface
         }
 
         //Unique customer found.
-        return $found->getItems()[0];
+        $items = $found->getItems();
+
+        return reset($items);
```
Comment (gabrieldagama) [inline app/code/Magento/Customer/Model/ForgotPasswordToken/GetCustomerByToken.php:89]: If the `$items` is an empty array `reset` function would return `false` and PHP would throw a fatal error, I guess it worth checking array length and throw an exception on such case, what do you think?

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @MTheProgrammer, looking at integration tests PHP `reset` function returns bool when the array provided is empty, can you have a look on it? Thanks!

**Comment by MTheProgrammer** [comment]
> Hi @MTheProgrammer, looking at integration tests PHP `reset` function returns bool when the array provided is empty, can you have a look on it? Thanks!

Haha, its missing throws at line 76

**Comment by MTheProgrammer** [comment]
@magento run all tests

**Comment by MTheProgrammer** [comment]
@magento run Functional Tests B2B

**Review by rogyar — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @rogyar, thank you for the review.
[ENGCOM-8150](https://jira.corp.magento.com/browse/ENGCOM-8150) has been created to process this Pull Request

**Review by engcom-Delta — CHANGES_REQUESTED** [review]
@MTheProgrammer  I am not able to reproduce issue descibed by PR on 2.4-develop
**### Manual testing scenarios:**
**Case 1:**
- Register customer
- Try to reset password from welcome email by clicking on `here` link (`Click here to reset it.`)
image
- Enter password and confirm it on `Set a New Password` page
- Click Set a New Password button
**Result:**
:heavy_check_mark: Password is changed and no errors in log

**Case 2:**
- Go to Customer Login page
- Click on `Forgot Your Password?` link
- Enter email and pass CAPTCHA
- Click Reset My Password button
- Click on Set A New Password link from email
- Enter password and confirm it on `Set a New Password` page
- Click Set a New Password button
**Result:**
:heavy_check_mark: Password is changed and no errors in log

**Case 3:**

- Create additional website `New Website`
image
- Set Add Store Code to Urls=Yes
- Set Share Customer Accounts=Per Website
- Register customer on Main Website
- Register customer on additional website with same email as on main website
- Go to Customer Login page on Main Website
- Click on `Forgot Your Password?` link
- Enter email and pass CAPTCHA
- Click Reset My Password button
- Click on Set A New Password link from email
- Enter password and confirm it on `Set a New Password` page
- Click Set a New Password button
- :heavy_check_mark: Password is changed and no errors in log
- Go to Customer Login page on New Website
- Click on `Forgot Your Password?` link
- Enter same email and pass CAPTCHA
- Click Reset My Password button
- Click on Set A New Password link from email
- Enter password and confirm it on `Set a New Password` page
- Click Set a New Password button
**Result:**
:heavy_check_mark: Password is changed and no errors in log

Could you provide exact scenario when customers are unable to reset their password?

**Comment by MTheProgrammer** [comment]
I don't have any exact scenario. Maybe Customer was trying to reset password twice. I don't know. Line 76 required a fix for sure where no exception was thrown and this issue also was discovered by tests in this PR.

**Comment by engcom-Charlie** [comment]
@MTheProgrammer we should find another way to reproduce the issue. Can you please help us?
Thank you.

**Comment by MTheProgrammer** [comment]
I see another MR fixed this issue: 
https://github.com/magento/magento2/issues/26288
My MR also adds unit tests whcih increase overall Magento code quality.

**Comment by engcom-Charlie** [comment]
@MTheProgrammer if the issue was fixed and covered by tests by https://github.com/magento/magento2/pull/27269 can I close this PR?

**Comment by MTheProgrammer** [comment]
You can, we need to verify whether other person is not working on the same issue already next time :)

**Pull request closed without merging**